File: spine-ios/Sources/Spine/SpineController.swift
import Foundation
import CoreGraphics
import QuartzCore
import UIKit

public typealias SpineControllerCallback = (_ controller: SpineController) -> Void

/// Controls how the skeleton of a ``SpineUIView`` is animated and rendered.
///
/// Upon initialization of a ``SpineUIView`` the provided `onInitialized` callback method is called once. This method can be used
/// to setup the initial animation(s) of the skeleton, among other things.
///
/// After initialization is complete, the ``SpineUIView`` is rendered at the screen refresh rate. In each frame,
/// the ``AnimationState`` is updated and applied to the ``Skeleton``.
///
/// Next the optionally provided method `onBeforeUpdateWorldTransforms` is called, which can modify the
/// skeleton before its current pose is calculated using ``Skeleton/updateWorldTransform(physics:)``. After
/// ``Skeleton.updateWorldTransforms`` has completed, the optional `onAfterUpdateWorldTransforms` method is
/// called, which can modify the current pose before rendering the skeleton.
///
/// Before the skeleton's current pose is rendered by the ``SpineUIView`` the optional `onBeforePaint` is called,
/// which allows rendering backgrounds or other objects that should go behind the skeleton in your view hierarchy. The
/// ``SpineUIView`` then renderes the skeleton's current pose, and finally calls the optional `onAfterPaint`, after which you
/// can render additional objects on top of the skeleton in your view hierarchy.
///
/// The underlying ``Atlas``, ``SkeletonData``, ``Skeleton``, ``AnimationStateData``, ``AnimationState``, and ``SkeletonDrawable``
/// can be accessed through their respective getters to inspect and/or modify the skeleton and its associated data. Accessing
/// this data is only allowed if the ``SpineUIView`` and its data have been initialized and have not been disposed yet.
///
/// By default, the view updates and renders the skeleton every frame. The `pause` method can be used to pause updating
/// and rendering the skeleton. The `resume` method resumes updating and rendering the skeleton. The `isPlaying` property
/// reports the current state.
///
/// Per default, ``SkeletonDrawableWrapper`` is disposed when ``SpineController`` is deinitialized. You can disable this behaviour with the ``disposeDrawableOnDeInit`` contructor parameter.
@objcMembers
public final class SpineController: NSObject, ObservableObject {
    
    public internal(set) var drawable: SkeletonDrawableWrapper!
    
    private let onInitialized: SpineControllerCallback?
    private let onBeforeUpdateWorldTransforms: SpineControllerCallback?
    private let onAfterUpdateWorldTransforms: SpineControllerCallback?
    private let onBeforePaint: SpineControllerCallback?
    private let onAfterPaint: SpineControllerCallback?
    private let disposeDrawableOnDeInit: Bool
    
    private var displayLink: CADisplayLink?
    
    private var scaleX: CGFloat = 1
    private var scaleY: CGFloat = 1
    private var offsetX: CGFloat = 0
    private var offsetY: CGFloat = 0
    
    @Published
    public private(set) var isPlaying: Bool = true
    
    @Published
    public private(set) var viewSize: CGSize = .zero
    
    /// Constructs a new ``SpineUIview`` controller. See the class documentation of ``SpineWidgetController`` for information on
    /// the optional arguments.
    public init(
        onInitialized: SpineControllerCallback? = nil,
        onBeforeUpdateWorldTransforms: SpineControllerCallback? = nil,
        onAfterUpdateWorldTransforms: SpineControllerCallback? = nil,
        onBeforePaint: SpineControllerCallback? = nil,
        onAfterPaint: SpineControllerCallback? = nil,
        disposeDrawableOnDeInit: Bool = true
    ) {
        self.onInitialized = onInitialized
        self.onBeforeUpdateWorldTransforms = onBeforeUpdateWorldTransforms
        self.onAfterUpdateWorldTransforms = onAfterUpdateWorldTransforms
        self.onBeforePaint = onBeforePaint
        self.onAfterPaint = onAfterPaint
        self.disposeDrawableOnDeInit = disposeDrawableOnDeInit
        
        super.init()
        
        addDisplayLinkIfNeeded()
    }
    
    deinit {
        if disposeDrawableOnDeInit {
            drawable?.dispose() // TODO move drawable out of view?
        }
        removeDisplayLink()
    }
    
    /// The ``Atlas`` from which images to render the skeleton are sourced.
    public var atlas: Atlas {
        drawable.atlas
    }
    
    /// The setup-pose data used by the skeleton.
    public var skeletonData: SkeletonData {
        drawable.skeletonData
    }
    
    /// The ``Skeleton``
    public var skeleton: Skeleton {
        drawable.skeleton
    }
    
    /// The mixing information used by the ``AnimationState``
    public var animationStateData: AnimationStateData {
        drawable.animationStateData
    }
    
    /// The ``AnimationState`` used to manage animations that are being applied to the
    /// skeleton.
    public var animationState: AnimationState {
        drawable.animationState
    }
    
    /// The ``AnimationStateWrapper`` used to hold ``AnimationState``, register ``AnimationStateListener`` and call ``AnimationStateWrapper/update(delta:)``
    public var animationStateWrapper: AnimationStateWrapper {
        drawable.animationStateWrapper
    }
    
    /// Transforms the coordinates given in the ``SpineUIView`` coordinate system in `position` to
    /// the skeleton coordinate system. See the `IKFollowing.swift` example how to use this
    /// to move a bone based on user touch input.
    public func toSkeletonCoordinates(position: CGPoint) -> CGPoint {
        let x = position.x;
        let y = position.y;
        return CGPoint(
            x: (x - viewSize.width / 2) / scaleX - offsetX,
            y: (y - viewSize.height / 2) / scaleY - offsetY
        )
    }
    
    /// Transforms the coordinates given in skeleton coordinate system to
    /// the the ``SpineUIView`` coordinates. See the `DebugRendering.swift` example hot to use this to draw rectangles over skeleton bones for debugging purposes.
    public func fromSkeletonCoordinates(position: CGPoint) -> CGPoint {
        let x = position.x;
        let y = position.y;
        return CGPoint(
            x: (x + offsetX) * scaleX,
            y: (y + offsetY) * scaleY
        )
    }
    
    /// Pauses updating and rendering the skeleton.
    public func pause() {
        isPlaying = false
    }
    
    /// Resumes updating and rendering the skeleton.
    public func resume() {
        isPlaying = true
    }
    
    internal func load(atlasFile: String, skeletonFile: String, bundle: Bundle = .main) async throws {
        let atlasAndPages = try await Atlas.fromBundle(atlasFile, bundle: bundle)
        let skeletonData = try await SkeletonData.fromBundle(
            atlas: atlasAndPages.0,
            skeletonFileName: skeletonFile,
            bundle: bundle
        )
        try await MainActor.run {
            let skeletonDrawableWrapper = try SkeletonDrawableWrapper(
                atlas: atlasAndPages.0,
                atlasPages: atlasAndPages.1,
                skeletonData: skeletonData
            )
            self.drawable = skeletonDrawableWrapper
        }
    }
    
    internal func initialize() {
        onInitialized?(self)
    }
    
    internal var lastUpdate: CFTimeInterval = 0
    
    @objc private func updateDrawable() {
        guard isPlaying else {
            lastUpdate = CACurrentMediaTime()
            return
        }
        
        if lastUpdate == 0 {
            lastUpdate = CACurrentMediaTime()
        }
        let delta = CACurrentMediaTime() - lastUpdate
        onBeforeUpdateWorldTransforms?(self)
        drawable?.update(delta: Float(delta))
        lastUpdate = CACurrentMediaTime()
        onAfterUpdateWorldTransforms?(self)
    }
    
    private func addDisplayLinkIfNeeded() {
        guard displayLink == nil else {
            return
        }
        displayLink = CADisplayLink(target: self, selector: #selector(updateDrawable))
        displayLink?.add(to: .current, forMode: .common)
    }

    private func removeDisplayLink() {
        displayLink?.remove(from: .current, forMode: .common)
        displayLink = nil
    }
}

extension SpineController: SpineRendererDelegate {
    
    func spineRendererWillDraw(_ spineRenderer: SpineRenderer) {
        onBeforePaint?(self)
    }
    
    func spineRendererDidDraw(_ spineRenderer: SpineRenderer) {
        onAfterPaint?(self)
    }
    
    func spineRendererDidUpdate(_ spineRenderer: SpineRenderer, scaleX: CGFloat, scaleY: CGFloat, offsetX: CGFloat, offsetY: CGFloat, size: CGSize) {
        self.scaleX = scaleX
        self.scaleY = scaleY
        self.offsetX = offsetX
        self.offsetY = offsetY
        self.viewSize = size
    }
}

extension SpineController: SpineRendererDataSource {
    
    func isPlaying(_ spineRenderer: SpineRenderer) -> Bool {
        return isPlaying
    }
    
    func skeletonDrawable(_ spineRenderer: SpineRenderer) -> SkeletonDrawableWrapper {
        return drawable
    }
    
    func renderCommands(_ spineRenderer: SpineRenderer) -> [RenderCommand] {
        return drawable?.skeletonDrawable.render() ?? []
    }
}

File: spine-android/spine-android/src/main/java/com/esotericsoftware/spine/android/SpineController.java
package com.esotericsoftware.spine.android;

import android.graphics.Canvas;
import android.graphics.Point;

import androidx.annotation.Nullable;

import com.badlogic.gdx.utils.Array;
import com.esotericsoftware.spine.AnimationState;
import com.esotericsoftware.spine.AnimationStateData;
import com.esotericsoftware.spine.Skeleton;
import com.esotericsoftware.spine.SkeletonData;
import com.esotericsoftware.spine.android.callbacks.SpineControllerAfterPaintCallback;
import com.esotericsoftware.spine.android.callbacks.SpineControllerBeforePaintCallback;
import com.esotericsoftware.spine.android.callbacks.SpineControllerCallback;

/**
 * Controls how the skeleton of a {@link SpineView} is animated and rendered.
 *
 * Upon initialization of a {@link SpineView}, the provided {@code onInitialized} callback method is called once. This method can be used
 * to set up the initial animation(s) of the skeleton, among other things.
 *
 * After initialization is complete, the {@link SpineView} is rendered at the screen refresh rate. In each frame,
 * the {@link AnimationState} is updated and applied to the {@link Skeleton}.
 *
 * Next, the optionally provided method {@code onBeforeUpdateWorldTransforms} is called, which can modify the
 * skeleton before its current pose is calculated using {@link Skeleton#updateWorldTransform(Skeleton.Physics)}. After
 * {@link Skeleton#updateWorldTransform(Skeleton.Physics)} has completed, the optional {@code onAfterUpdateWorldTransforms} method is
 * called, which can modify the current pose before rendering the skeleton.
 *
 * Before the skeleton's current pose is rendered by the {@link SpineView}, the optional {@code onBeforePaint} is called,
 * which allows rendering backgrounds or other objects that should go behind the skeleton on the {@link Canvas}. The
 * {@link SpineView} then renders the skeleton's current pose and finally calls the optional {@code onAfterPaint}, which
 * can render additional objects on top of the skeleton.
 *
 * The underlying {@link AndroidTextureAtlas}, {@link SkeletonData}, {@link Skeleton}, {@link AnimationStateData}, {@link AnimationState}, and {@link AndroidSkeletonDrawable}
 * can be accessed through their respective getters to inspect and/or modify the skeleton and its associated data. Accessing
 * this data is only allowed if the {@link SpineView} and its data have been initialized and have not been disposed of yet.
 *
 * By default, the widget updates and renders the skeleton every frame. The {@code pause} method can be used to pause updating
 * and rendering the skeleton. The {@link SpineController#resume()} method resumes updating and rendering the skeleton. The {@link SpineController#isPlaying()} getter
 * reports the current state.
 */
public class SpineController {
    /**
     * Used to build {@link SpineController} instances.
     * */
    public static class Builder {
        private final SpineControllerCallback onInitialized;
        private SpineControllerCallback onBeforeUpdateWorldTransforms;
        private SpineControllerCallback onAfterUpdateWorldTransforms;
        private SpineControllerBeforePaintCallback onBeforePaint;
        private SpineControllerAfterPaintCallback onAfterPaint;

        /**
         * Instantiate a {@link Builder} used to build a {@link SpineController}, which controls how the skeleton of a {@link SpineView}
         * is animated and rendered. Upon initialization of a {@link SpineView}, the provided {@code onInitialized} callback
         * method is called once. This method can be used to set up the initial animation(s) of the skeleton, among other things.
         *
         * @param onInitialized Upon initialization of a {@link SpineView}, the provided {@code onInitialized} callback
         *                      method is called once. This method can be used to set up the initial animation(s) of the skeleton,
         *                      among other things.
         */
        public Builder(SpineControllerCallback onInitialized) {
            this.onInitialized = onInitialized;
        }

        /**
         * Sets the {@code onBeforeUpdateWorldTransforms} callback. It is called before the skeleton's current pose is calculated
         * using {@link Skeleton#updateWorldTransform(Skeleton.Physics)}. It can be used to modify the skeleton before the pose calculation.
         */
        public Builder setOnBeforeUpdateWorldTransforms(SpineControllerCallback onBeforeUpdateWorldTransforms) {
            this.onBeforeUpdateWorldTransforms = onBeforeUpdateWorldTransforms;
            return this;
        }

        /**
         * Sets the {@code onAfterUpdateWorldTransforms} callback. This method is called after the skeleton's current pose is calculated using
         * {@link Skeleton#updateWorldTransform(Skeleton.Physics)}. It can be used to modify the current pose before rendering the skeleton.
         */
        public Builder setOnAfterUpdateWorldTransforms(SpineControllerCallback onAfterUpdateWorldTransforms) {
            this.onAfterUpdateWorldTransforms = onAfterUpdateWorldTransforms;
            return this;
        }

        /**
         * Sets the {@code onBeforePaint} callback. It is called before the skeleton's current pose is rendered by the
         * {@link SpineView}. It allows rendering backgrounds or other objects that should go behind the skeleton on the
         * {@link Canvas}.
         */
        public Builder setOnBeforePaint(SpineControllerBeforePaintCallback onBeforePaint) {
            this.onBeforePaint = onBeforePaint;
            return this;
        }

        /**
         * Sets the {@code onAfterPaint} callback. It is called after the skeleton's current pose is rendered by the
         * {@link SpineView}. It allows rendering additional objects on top of the skeleton.
         */
        public Builder setOnAfterPaint(SpineControllerAfterPaintCallback onAfterPaint) {
            this.onAfterPaint = onAfterPaint;
            return this;
        }

        public SpineController build() {
            SpineController spineController = new SpineController(onInitialized);
            spineController.onBeforeUpdateWorldTransforms = onBeforeUpdateWorldTransforms;
            spineController.onAfterUpdateWorldTransforms = onAfterUpdateWorldTransforms;
            spineController.onBeforePaint = onBeforePaint;
            spineController.onAfterPaint = onAfterPaint;
            return spineController;
        }
    }

    private final SpineControllerCallback onInitialized;
    private @Nullable SpineControllerCallback onBeforeUpdateWorldTransforms;
    private @Nullable SpineControllerCallback onAfterUpdateWorldTransforms;
    private @Nullable SpineControllerBeforePaintCallback onBeforePaint;
    private @Nullable SpineControllerAfterPaintCallback onAfterPaint;
    private AndroidSkeletonDrawable drawable;
    private boolean playing = true;
    private double offsetX = 0;
    private double offsetY = 0;
    private double scaleX = 1;
    private double scaleY = 1;

    /**
     * Instantiate a {@link SpineController}, which controls how the skeleton of a {@link SpineView} is animated and rendered.
     * Upon initialization of a {@link SpineView}, the provided {@code onInitialized} callback method is called once.
     * This method can be used to set up the initial animation(s) of the skeleton, among other things.
     *
     * @param onInitialized Upon initialization of a {@link SpineView}, the provided {@code onInitialized} callback
     *                      method is called once. This method can be used to set up the initial animation(s) of the skeleton,
     *                      among other things.
     */
    public SpineController(SpineControllerCallback onInitialized) {
        this.onInitialized = onInitialized;
    }

    protected void init(AndroidSkeletonDrawable drawable) {
        this.drawable = drawable;
        if (onInitialized != null) {
            onInitialized.execute(this);
        }
    }

    /**
     * The {@link AndroidTextureAtlas} from which images to render the skeleton are sourced.
     */
    public AndroidTextureAtlas getAtlas() {
        if (drawable == null) throw new RuntimeException("Controller is not initialized yet.");
        return drawable.getAtlas();
    }

    /**
     * The setup-pose data used by the skeleton.
     */
    public SkeletonData getSkeletonDate() {
        if (drawable == null) throw new RuntimeException("Controller is not initialized yet.");
        return drawable.getSkeletonData();
    }

    /**
     * The {@link Skeleton}.
     */
    public Skeleton getSkeleton() {
        if (drawable == null) throw new RuntimeException("Controller is not initialized yet.");
        return drawable.getSkeleton();
    }

    /**
     * The mixing information used by the {@link AnimationState}.
     */
    public AnimationStateData getAnimationStateData() {
        if (drawable == null) throw new RuntimeException("Controller is not initialized yet.");
        return drawable.getAnimationStateData();
    }

    /**
     * The {@link AnimationState} used to manage animations that are being applied to the
     * skeleton.
     */
    public AnimationState getAnimationState() {
        if (drawable == null) throw new RuntimeException("Controller is not initialized yet.");
        return drawable.getAnimationState();
    }

    /**
     * The {@link AndroidSkeletonDrawable}.
     */
    public AndroidSkeletonDrawable getDrawable() {
        if (drawable == null) throw new RuntimeException("Controller is not initialized yet.");
        return drawable;
    }

    /**
     * Checks if the {@link  SpineView} is initialized.
     */
    public boolean isInitialized() {
        return drawable != null;
    }

    /**
     * Checks if the animation is currently playing.
     */
    public boolean isPlaying() {
        return playing;
    }

    /**
     * Pauses updating and rendering the skeleton.
     */
    public void pause() {
        if (playing) {
            playing = false;
        }
    }

    /**
     * Resumes updating and rendering the skeleton.
     */
    public void resume() {
        if (!playing) {
            playing = true;
        }
    }

    /**
     * Transforms the coordinates given in the {@link SpineView} coordinate system in {@code position} to
     * the skeleton coordinate system. See the {@code IKFollowing.kt} example for how to use this
     * to move a bone based on user touch input.
     */
    public Point toSkeletonCoordinates(Point position) {
        int x = position.x;
        int y = position.y;
        return new Point((int) (x / scaleX - offsetX), (int) (y / scaleY - offsetY));
    }

    /**
     * Sets the {@code onBeforeUpdateWorldTransforms} callback. It is called before the skeleton's current pose is calculated
     * using {@link Skeleton#updateWorldTransform(Skeleton.Physics)}. It can be used to modify the skeleton before the pose calculation.
     */
    public void setOnBeforeUpdateWorldTransforms(@Nullable SpineControllerCallback onBeforeUpdateWorldTransforms) {
        this.onBeforeUpdateWorldTransforms = onBeforeUpdateWorldTransforms;
    }

    /**
     * Sets the {@code onAfterUpdateWorldTransforms} callback. This method is called after the skeleton's current pose is calculated using
     * {@link Skeleton#updateWorldTransform(Skeleton.Physics)}. It can be used to modify the current pose before rendering the skeleton.
     */
    public void setOnAfterUpdateWorldTransforms(@Nullable SpineControllerCallback onAfterUpdateWorldTransforms) {
        this.onAfterUpdateWorldTransforms = onAfterUpdateWorldTransforms;
    }

    /**
     * Sets the {@code onBeforePaint} callback. It is called before the skeleton's current pose is rendered by the
     * {@link SpineView}. It allows rendering backgrounds or other objects that should go behind the skeleton on the
     * {@link Canvas}.
     */
    public void setOnBeforePaint(@Nullable SpineControllerBeforePaintCallback onBeforePaint) {
        this.onBeforePaint = onBeforePaint;
    }

    /**
     * Sets the {@code onAfterPaint} callback. It is called after the skeleton's current pose is rendered by the
     * {@link SpineView}. It allows rendering additional objects on top of the skeleton.
     */
    public void setOnAfterPaint(@Nullable SpineControllerAfterPaintCallback onAfterPaint) {
        this.onAfterPaint = onAfterPaint;
    }

    protected void setCoordinateTransform(double offsetX, double offsetY, double scaleX, double scaleY) {
        this.offsetX = offsetX;
        this.offsetY = offsetY;
        this.scaleX = scaleX;
        this.scaleY = scaleY;
    }

    protected void callOnBeforeUpdateWorldTransforms() {
        if (onBeforeUpdateWorldTransforms != null) {
            onBeforeUpdateWorldTransforms.execute(this);
        }
    }

    protected void callOnAfterUpdateWorldTransforms() {
        if (onAfterUpdateWorldTransforms != null) {
            onAfterUpdateWorldTransforms.execute(this);
        }
    }

    protected void callOnBeforePaint(Canvas canvas) {
        if (onBeforePaint != null) {
            onBeforePaint.execute(this, canvas);
        }
    }

    protected void callOnAfterPaint(Canvas canvas, Array<SkeletonRenderer.RenderCommand> renderCommands) {
        if (onAfterPaint != null) {
            onAfterPaint.execute(this, canvas, renderCommands);
        }
    }
}
